src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs

```diff
@@ -0,0 +1,165 @@
+        {
+            if (symbol is null || useSiteInfo.IsDiscarded || !(addDiagnostics || addDependencies))
+            {
+                return;
+            }
+
+            var info = symbol.GetUseSiteInfo();
+
+            if (addDiagnostics)
+            {
+                useSiteInfo.AddDiagnostics(info);
+            }
+            else
+            {
+                Debug.Assert(info.DiagnosticInfo == null || addDependencies); // It would be strange to drop diagnostics, but record dependencies
+            }
+
+            if (addDependencies)
+            {
+                useSiteInfo.AddDependencies(info);
+            }
+        }
+    }
+
+    internal sealed class BindingDiagnosticBag : BindingDiagnosticBag<AssemblySymbol>
+    {
+        public static readonly BindingDiagnosticBag Discarded = new BindingDiagnosticBag(null, null);
+
+        public BindingDiagnosticBag()
+            : this(usePool: false)
+        { }
+
+        private BindingDiagnosticBag(bool usePool)
+            : base(usePool)
+        { }
+
+        public BindingDiagnosticBag(DiagnosticBag? diagnosticBag)
+            : base(diagnosticBag, dependenciesBag: null)
+        {
+        }
+
+        public BindingDiagnosticBag(DiagnosticBag? diagnosticBag, ICollection<AssemblySymbol>? dependenciesBag)
+            : base(diagnosticBag, dependenciesBag)
+        {
+        }
+
+        internal static BindingDiagnosticBag GetInstance()
+        {
+            return new BindingDiagnosticBag(usePool: true);
+        }
+
+        // PROTOTYPE(UsedAssemblyReferences): Remove this overload once conversion operators are removed from the base type.
+        internal void AddRange(BindingDiagnosticBag other)
+        {
+            AddRange((BindingDiagnosticBag<AssemblySymbol>)other);
+        }
+
+        internal void AddDependencies(Symbol? symbol)
+        {
+            if (symbol is object && DependenciesBag is object)
+            {
+                AddDependencies(symbol.GetUseSiteInfo());
+            }
+        }
+
+        internal bool ReportUseSite(Symbol? symbol, SyntaxNode node)
+        {
+            return ReportUseSite(symbol, node.Location);
+        }
+
+        internal bool ReportUseSite(Symbol? symbol, SyntaxToken token)
+        {
+            return ReportUseSite(symbol, token.GetLocation());
+        }
+
+        internal bool ReportUseSite(Symbol? symbol, Location location)
+        {
+            if (symbol is object)
+            {
+                return Add(symbol.GetUseSiteInfo(), location);
+            }
+
+            return false;
+        }
+
+        internal void AddAssembliesUsedByNamespaceReference(NamespaceSymbol ns)
+        {
+            if (DependenciesBag is null)
+            {
+                return;
+            }
+
+            addAssembliesUsedByNamespaceReference(ns);
+
+            void addAssembliesUsedByNamespaceReference(NamespaceSymbol ns)
```

Comment:
>addAssembliesUsedByNamespaceReference [](start = 17, length = 37)

Given that this is only used in one place and recursively calls the original function, which only differs by case of the first letter, please inline to reduce confusion.

Reply:
Or, if this was supposed to recursively call the local function, please rename to something like `assAssembliesUsedByNamespaceReferenceRecursive` to make the intention obvious.

---
In reply to: [361554930](https://github.com/dotnet/roslyn/pull/40560#discussion_r361554930) [](ancestors = 361554930)

Reply:
This function is recursive. We don.t have a requirement to reflect recursive nature of functions in their names.  

---
In reply to: [361555026](https://github.com/dotnet/roslyn/pull/40560#discussion_r361555026) [](ancestors = 361555026,361554930)

Reply:
> We don.t have a requirement to reflect recursive nature of functions in their names.

The point isn't to reflect the recursive nature, the point is to be explicit about your intent. The correctness problem occurred because two functions only differed by one letter, which makes it really easy to misread and/or mistype.

---
In reply to: [361555546](https://github.com/dotnet/roslyn/pull/40560#discussion_r361555546) [](ancestors = 361555546,361555026,361554930)